### PR TITLE
Fixed bug that prevents retrieval of IPAddress using its address args if it was created using host and prefix_length.

### PR DESCRIPTION
**Pull request by timizuoebideri1**
# Closes: #4733 
# What's Changed

# TODO

- [ ] Explanation of Change(s)
- [ ] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/#creating-changelog-fragments))
- [ ] Attached Screenshots, Payload Example
- [ ] Unit, Integration Tests
- [ ] Documentation Updates (when adding/changing features)
- [ ] Example Plugin Updates (when adding/changing features)
- [ ] Outline Remaining Work, Constraints from Design

### CODE REVIEW

File: nautobot/ipam/signals.py [hunk 1]
```diff
@@ -0,0 +1,13 @@
+from django.db.models.signals import pre_save
+from django.dispatch import receiver
+
+from nautobot.ipam.models import IPAddress
+
+
+@receiver(pre_save, sender=IPAddress)
```
Comment (glennmatthews) [inline nautobot/ipam/signals.py:7]: Why does this need to be done as a signal, instead of directly within `IPAddress.save()`?

**Comment by gsnider2195** [comment]
Should we fix VarbinaryIPField as part of this? The root of this issue is VarbinaryIPField is creating a row with a broadcast of "::" when no broadcast is provided in the ORM. The broadcast field is not allowed to be null so when you don't provide it the database should raise a not-null constraint error.

**Comment by gsnider2195** [comment]
```py
>>> IPAddress.objects.create(status=s, prefix_length=32, host="10.0.0.0")
```
```sql
INSERT INTO "ipam_ipaddress" (
    "id",
    "created",
    "last_updated",
    "_custom_field_data",
    "status_id",
    "host",
    "broadcast",
    "prefix_length",
    "vrf_id",
    "tenant_id",
    "role",
    "assigned_object_type_id",
    "assigned_object_id",
    "nat_inside_id",
    "dns_name",
    "description"
    ) VALUES (
        'b79f5924-7ada-4d9b-8a9d-aef3a2d955cb'::uuid,
        '2023-11-08'::date,
        '2023-11-08T20:47:01.164629+00:00'::timestamptz,
        '{}',
        'c119032d-b03f-46aa-9f91-8996a1716abe'::uuid,
        '\\x0a000000'::bytea,
        '\\x00000000000000000000000000000000'::bytea,
        32,
        NULL,
        NULL,
        '',
        NULL,
        NULL,
        NULL,
        '',
        ''
    )
```

The null broadcast is being inserted as `'\\x00000000000000000000000000000000'::bytea`

**Comment by gsnider2195** [comment]
This seems to work:

```diff
diff --git a/nautobot/ipam/fields.py b/nautobot/ipam/fields.py
index c5e10144b..292f4864f 100644
--- a/nautobot/ipam/fields.py
+++ b/nautobot/ipam/fields.py
@@ -68,8 +68,8 @@ class VarbinaryIPField(models.BinaryField):
 
     def get_db_prep_value(self, value, connection, prepared=False):
         """Converts Python (str) to DB (varbinary)."""
-        if value is None:
-            return value
+        if value in self.empty_values:
+            return None
 
         # Parse the address and then pack it to binary.
         value = self._parse_address(value).packed
```

```py
>>> IPAddress.objects.create(status=s, prefix_length=32, host="10.0.0.0")
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
  File "/usr/local/lib/python3.10/site-packages/django_prometheus/db/common.py", line 69, in execute
    return super().execute(*args, **kwargs)
psycopg2.errors.NotNullViolation: null value in column "broadcast" of relation "ipam_ipaddress" violates not-null constraint
DETAIL:  Failing row contains (bf5ed771-2429-4b9d-84f6-cceb378d105c, 2023-11-08, 2023-11-08 21:18:22.804999+00, {}, \x0a000000, null, 32, , null, , , null, null, c119032d-b03f-46aa-9f91-8996a1716abe, null, null).


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/usr/local/lib/python3.10/site-packages/django/db/models/manager.py", line 85, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/django/db/models/query.py", line 453, in create
    obj.save(force_insert=True, using=self.db)
  File "/usr/local/lib/python3.10/site-packages/django/db/models/base.py", line 739, in save
    self.save_base(using=using, force_insert=force_insert,
  File "/usr/local/lib/python3.10/site-packages/django/db/models/base.py", line 776, in save_base
    updated = self._save_table(
  File "/usr/local/lib/python3.10/site-packages/django/db/models/base.py", line 881, in _save_table
    results = self._do_insert(cls._base_manager, using, fields, returning_fields, raw)
  File "/usr/local/lib/python3.10/site-packages/django/db/models/base.py", line 919, in _do_insert
    return manager._insert(
  File "/usr/local/lib/python3.10/site-packages/django/db/models/manager.py", line 85, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/django/db/models/query.py", line 1270, in _insert
    return query.get_compiler(using=using).execute_sql(returning_fields)
  File "/usr/local/lib/python3.10/site-packages/django/db/models/sql/compiler.py", line 1416, in execute_sql
    cursor.execute(sql, params)
  File "/usr/local/lib/python3.10/site-packages/django/db/backends/utils.py", line 98, in execute
    return super().execute(sql, params)
  File "/usr/local/lib/python3.10/site-packages/cacheops/transaction.py", line 98, in execute
    result = self._no_monkey.execute(self, sql, params)
  File "/usr/local/lib/python3.10/site-packages/django/db/backends/utils.py", line 66, in execute
    return self._execute_with_wrappers(sql, params, many=False, executor=self._execute)
  File "/usr/local/lib/python3.10/site-packages/django/db/backends/utils.py", line 75, in _execute_with_wrappers
    return executor(sql, params, many, context)
  File "/usr/local/lib/python3.10/site-packages/django/db/backends/utils.py", line 79, in _execute
    with self.db.wrap_database_errors:
  File "/usr/local/lib/python3.10/site-packages/django/db/utils.py", line 90, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "/usr/local/lib/python3.10/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
  File "/usr/local/lib/python3.10/site-packages/django_prometheus/db/common.py", line 69, in execute
    return super().execute(*args, **kwargs)
django.db.utils.IntegrityError: null value in column "broadcast" of relation "ipam_ipaddress" violates not-null constraint
DETAIL:  Failing row contains (bf5ed771-2429-4b9d-84f6-cceb378d105c, 2023-11-08, 2023-11-08 21:18:22.804999+00, {}, \x0a000000, null, 32, , null, , , null, null, c119032d-b03f-46aa-9f91-8996a1716abe, null, null).
```

**Comment by timizuoebideri1** [comment]
> ```python
> >>> IPAddress.objects.create(status=s, prefix_length=32, host="10.0.0.0")
> ```
> 
>  ```sql
> INSERT INTO "ipam_ipaddress" (
>     "id",
>     "created",
>     "last_updated",
>     "_custom_field_data",
>     "status_id",
>     "host",
>     "broadcast",
>     "prefix_length",
>     "vrf_id",
>     "tenant_id",
>     "role",
>     "assigned_object_type_id",
>     "assigned_object_id",
>     "nat_inside_id",
>     "dns_name",
>     "description"
>     ) VALUES (
>         'b79f5924-7ada-4d9b-8a9d-aef3a2d955cb'::uuid,
>         '2023-11-08'::date,
>         '2023-11-08T20:47:01.164629+00:00'::timestamptz,
>         '{}',
>         'c119032d-b03f-46aa-9f91-8996a1716abe'::uuid,
>         '\\x0a000000'::bytea,
>         '\\x00000000000000000000000000000000'::bytea,
>         32,
>         NULL,
>         NULL,
>         '',
>         NULL,
>         NULL,
>         NULL,
>         '',
>         ''
>     )
> ```
> 
> The null broadcast is being inserted as `'\\x00000000000000000000000000000000'::bytea`

But, if we already derive the 'broadcast' for other use cases other than ORM '.create()', should we not stay to this pattern?

**Comment by gsnider2195** [comment]
> > ```python
> > >>> IPAddress.objects.create(status=s, prefix_length=32, host="10.0.0.0")
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ```sql
> > INSERT INTO "ipam_ipaddress" (
> >    "id",
> >    "created",
> >    "last_updated",
> >    "_custom_field_data",
> >    "status_id",
> >    "host",
> >    "broadcast",
> >    "prefix_length",
> >    "vrf_id",
> >    "tenant_id",
> >    "role",
> >    "assigned_object_type_id",
> >    "assigned_object_id",
> >    "nat_inside_id",
> >    "dns_name",
> >    "description"
> >    ) VALUES (
> >        'b79f5924-7ada-4d9b-8a9d-aef3a2d955cb'::uuid,
> >        '2023-11-08'::date,
> >        '2023-11-08T20:47:01.164629+00:00'::timestamptz,
> >        '{}',
> >        'c119032d-b03f-46aa-9f91-8996a1716abe'::uuid,
> >        '\\x0a000000'::bytea,
> >        '\\x00000000000000000000000000000000'::bytea,
> >        32,
> >        NULL,
> >        NULL,
> >        '',
> >        NULL,
> >        NULL,
> >        NULL,
> >        '',
> >        ''
> >    )
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The null broadcast is being inserted as `'\\x00000000000000000000000000000000'::bytea`
> 
> But, if we already derive the 'broadcast' for other use cases other than ORM '.create()', should we not stay to this pattern?

I do think that we should continue to generate the field but we should also fix the VarbinaryField. Perhaps I can just open a separate issue for that. We can talk about it today

File: nautobot/ipam/tests/test_models.py [hunk 2]
```diff
@@ -381,15 +381,27 @@ def test_duplicate_vrf_unique(self):
...
+        ipaddress = IPAddress.objects.create(host="10.35.1.251", prefix_length=32, status=self.status)
         self.assertTrue(IPAddress.objects.filter(address=str(ipaddress)).exists())
 
+    def test_varbinary_ip_fields_with_empty_values_do_not_violate_not_null_constrains(self):
+        # Assert that an error is triggered when the host is not provided.
+        # Initially, VarbinaryIPField fields with None values are stored as the binary representation of b'',
+        # thereby bypassing the Not Null Constraint check.
+        with self.assertRaises(IntegrityError) as err:
+            IPAddress.objects.create(prefix_length=32, status=self.status)
+        self.assertIn(
```
Comment (gsnider2195) [inline nautobot/ipam/tests/test_models.py:400]: Is this the same error message in postgres and mysql?

Comment (glennmatthews) [reply nautobot/ipam/tests/test_models.py:400]:
```
 AssertionError: 'null value in column "host" of relation "ipam_ipaddress" violates not-null constraint' not found in '(1048, "Column \'host\' cannot be null")'
```

File: nautobot/ipam/tests/test_models.py [hunk 3]
```diff
@@ -381,6 +381,27 @@ def test_duplicate_vrf_unique(self):
...
+            'null value in column "host" of relation "ipam_ipaddress" violates not-null constraint',
+            str(err.exception),
+        )
```
Comment (timizuoebideri1) [inline nautobot/ipam/tests/test_models.py:403]:
```suggestion
```

File: nautobot/ipam/fields.py [hunk 4]
```diff
@@ -23,6 +23,13 @@ def db_type(self, connection):
         # Or 'varbinary' for everyone else.
         return "varbinary(16)"
 
+    def get_default(self):
+        value = super().get_default()
+        # Prevent None or "" values from been represented as b''
```
Comment (gsnider2195) [inline nautobot/ipam/fields.py:28]:
```suggestion
        # Prevent None or "" values from being represented as b''
```

File: nautobot/ipam/tests/test_models.py [hunk 5]
```diff
@@ -381,6 +381,23 @@ def test_duplicate_vrf_unique(self):
...
+        """Assert bug https://github.com/nautobot/nautobot/issues/4733 Unable to retrieve IPAddress using `address`
+        args if it was created using host and prefix_length is resolved"""
+        ipaddress = IPAddress.objects.create(host="10.35.1.251", prefix_length=32, status=self.status)
+        self.assertTrue(IPAddress.objects.filter(address=str(ipaddress)).exists())
+
+    def test_varbinary_ip_fields_with_empty_values_do_not_violate_not_null_constrains(self):
+        # Assert that an error is triggered when the host is not provided.
+        # Initially, VarbinaryIPField fields with None values are stored as the binary representation of b'',
+        # thereby bypassing the Not Null Constraint check.
+        with self.assertRaises(IntegrityError) as err:
```
Comment (gsnider2195) [inline nautobot/ipam/tests/test_models.py:398]:
```suggestion
        with self.assertRaises(IntegrityError):
```

**Comment by glennmatthews** [comment]
I actually hadn't noticed that this PR was against `ltm-1.6` rather than `develop` until now. Bug fixes should first go into `develop` and only be backported to LTM if there's a significant need.

**Comment by gsnider2195** [comment]
> I actually hadn't noticed that this PR was against `ltm-1.6` rather than `develop` until now. Bug fixes should first go into `develop` and only be backported to LTM if there's a significant need.

I think the original bug only affects 1.6. The VarbinaryField fix affects both.

**Comment by glennmatthews** [comment]
OK, let's verify that the original bug is specific to 1.6 and get a PR opened with the VarBinaryField fix.

**Review by glennmatthews — APPROVED** [review]

**Review by gsnider2195 — APPROVED** [review]

**Pull request merged**